cpp/include/cudf/strings/detail/gather.cuh

```diff
@@ -34,6 +34,139 @@ namespace cudf {
+
+  regs.x = __funnelshift_r(regs.x, regs.y, shift);
+  regs.y = __funnelshift_r(regs.y, regs.z, shift);
+  regs.z = __funnelshift_r(regs.z, regs.w, shift);
+  regs.w = __funnelshift_r(regs.w, tail, shift);
+
+  return regs;
+}
+
+template <typename StringIterator, typename MapIterator>
+__global__ void gather_chars_fn_string_parallel(StringIterator strings_begin,
+                                                char* out_chars,
+                                                cudf::device_span<int32_t const> const out_offsets,
+                                                MapIterator string_indices,
+                                                size_type total_out_strings)
+{
+  int global_thread_id = blockIdx.x * blockDim.x + threadIdx.x;
+  int global_warp_id   = global_thread_id / 32;
+  int warp_lane        = global_thread_id % 32;
+  int nwarps           = gridDim.x * blockDim.x / 32;
+
+  size_t alignment_offset  = reinterpret_cast<size_t>(out_chars) & 15;
+  uint4* out_chars_aligned = reinterpret_cast<uint4*>(out_chars - alignment_offset);
+
+  for (size_type istring = global_warp_id; istring < total_out_strings; istring += nwarps) {
+    auto out_start = out_offsets[istring];
+    auto out_end   = out_offsets[istring + 1];
+
+    // This check is necessary because string_indices[istring] may be out of bound.
+    if (out_start == out_end) continue;
+
+    const char* in_start = strings_begin[string_indices[istring]].data();
+
+    int32_t out_start_aligned = (out_start + alignment_offset + 15) / 16 * 16 - alignment_offset;
+    int32_t out_end_aligned   = (out_end + alignment_offset) / 16 * 16 - alignment_offset;
+
+    for (size_type ichar = out_start_aligned + warp_lane * 16; ichar < out_end_aligned;
+         ichar += 32 * 16) {
+      *(out_chars_aligned + (ichar + alignment_offset) / 16) =
+        load_uint4(in_start + ichar - out_start);
+    }
+
+    if (out_end_aligned <= out_start_aligned) {
+      int32_t ichar = out_start + warp_lane;
+      if (ichar < out_end) { out_chars[ichar] = in_start[warp_lane]; }
+    } else {
+      if (out_start + warp_lane < out_start_aligned) {
+        out_chars[out_start + warp_lane] = in_start[warp_lane];
+      }
+
+      int32_t ichar = out_end_aligned + warp_lane;
+      if (ichar < out_end) { out_chars[ichar] = in_start[ichar - out_start]; }
+    }
+  }
+}
+
+// Strategy 2: Char-parallel
+// This strategy assigns characters to threads, and uses binary search for getting the string
+// index. To improve the binary search performance, fixed number of strings per threadblock is
+// used. This strategy is best suited for small strings.
+constexpr static int strings_per_threadblock = 32;
+
+// Binary search `value` in `offsets` of length `nelements`. Require `nelements` to be less than or
+// equal to `strings_per_threadblock`. Require `strings_per_threadblock` to be an exponential of 2.
+__forceinline__ __device__ size_type binary_search(int32_t* offsets,
```

Comment:
Can't this just be a `thrust::lower_bound` call instead?

Reply:
My motivation for implementing this myself is that since `strings_per_threadblock` is known at compile time, the compiler can unroll the binary search loop.

I will investigate the performance impact of using `thrust::lower_bound`.

Reply:
I tried `thrust::upper_bound`, and the overall performance difference is about 2%. I think it's probably not worthwhile for us to maintain a custom kernel for 2% performance.